### PR TITLE
Incorporate cadre allocation computations more into mainline

### PR DESCRIPTION
**Pull request by celiot-IDM**
Phew ... this work started out with a small idea and turned into something very big. 

The core work takes the cadre allocation computations that used to be done by functions in pace_post_process.R and moves them more into the mainline of process. So instead of using suite results CSV files as inputs, post-processing information can now be done immediately after an experiment finishes, based on the `results` structure returned by `RunExperiments()`.

NOTE: THIS IS A BREAKING CHANGE. Old versions of the post-processing code won't work with the new Cadre allocation sheet format. A side-effect of making cadre roles user-configurable at the scenario/delivery-model level is that we can't naively combine suite result CSV files the way it was done before by `ReadAndCollateSuiteResults()`. (The key problem is that a user could configure a member ID - e.g "HEW1" - to mean one thing in one delivery model, but something different in another delivery model. Different delivery models have different cadre memberships - MergedModel = {HEW1 | HEW2 | FHP1 | FHP2}, but BasicModel = {HEW1 | HEW2 | FHP1 | FHP2 | RN | EHP} - so it's possible that in trying to create one big view that crosses several delivery models, we could end up aggregating things that shouldn't be aggregated.)

Annual overheads are calculated for all cadre members in all models, but I'm not sure how you want to access that information. 

Here's a sample of the new approach. 

```
library(pacehrh)
pacehrh::Trace(TRUE)

pacehrh::InitializePopulation()
pacehrh::InitializeScenarios()
pacehrh::InitializeStochasticParameters()
pacehrh::InitializeSeasonality()
pacehrh::InitializeCadreRoles()

pacehrh::SetGlobalStartEndYears(2020, 2050)
pacehrh::SetRoundingLaw("none")

results <-
  pacehrh::RunExperiments(scenarioName = "MergedModel",
                          trials = 20)

SR <- pacehrh::SaveExtendedSuiteResults(results, run = "Run_ID")
CA <- pacehrh::SaveCadreAllocations(SR)

summaryStats <- pacehrh::ComputeSummaryStats(SR, CA)
```
Two additional bits of magic ...

Magic 1: If you don't need to keep the output from `SaveExtendedSuiteResults()` (basically a much wider version of the standard output, but extended with Scenario and Task information), you can speed things up a bit by piping:

```
# Create the cadre allocation table without saving intermediate results locally or to disk
CA <-
  pacehrh::SaveExtendedSuiteResults(results) %>%
  pacehrh::SaveCadreAllocations()
```

Magic 2: Both `SaveExtendedSuiteResults()` and `SaveCadreAllocations()` take filepath parameters which will save their outputs to disk.
```
# Create CA table and save both the extended suite results and the cadre allocation results
CA <-
  pacehrh::SaveExtendedSuiteResults(results, filepath = "_SR.csv", run = "Run-100") %>%
  pacehrh::SaveCadreAllocations(filepath = "_CA.csv")
```

I have some more tidying up to do, but I'm going to move on to finish off the stochastic rate capping work now.

Addresses #51

### CODE REVIEW

**Comment by celiot-IDM** [comment]
@MeWu-IDM : Integration tests are failing, but I don't understand why. Over to you!

File: pacehrh/R/pace_cadre_process.R [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+#'
+#' @examples
+#' \dontrun{
+#' }
+SaveCadreAllocations <- function(suiteResults, filepath = NULL, annual = TRUE) {
+  scenarioName <- BVE$scenario$UniqueID
+
+  # Do some sanity checking
+  if (!identical(GPE$years, unique(suiteResults$Year))) {
+    traceMessage("Configured year range (GPE) does not match reported year range")
```
Comment (MeWu-IDM) [inline pacehrh/R/pace_cadre_process.R:194]: I ran into issue when loading suiteResults from a file where "Year" was read as integer and GPE$year is double, please cast them both to integer for identical comparison

**Comment by MeWu-IDM** [comment]
does pacehrh:::SaveCadreAllocations not work with the results from pacehrh::ReadAndCollateSuiteResults after this change?
I ran into error: more than one scenario in suite results

File: pacehrh/R/pace_cadre_process.R [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+                          by = dplyr::join_by(Task_ID == Indicator))
+
+  # Create a stochastic hours per week value that's different for each year.
+  sdvalue <-
+    BVE$stochasticParams$p[BVE$stochasticParams$Value == "Hours per week"]
+
+  # TODO: find a more efficient calculation.
+  stochasticHrs <- out[, .(Count = .N), by = .(Trial_num, Year)]
+
+  stochasticHrs$HPW_stochastic <-
```
Comment (MeWu-IDM) [inline pacehrh/R/pace_cadre_process.R:146]: We need documentation to explain why do we need to calculate this here, also from design perspective,  it seems odd to do calculation in "save results" methods.

Comment (celiot-IDM) [reply pacehrh/R/pace_cadre_process.R:146]: This was in the original post-processing, so I kept it. But I agree, it doesn't make a lot of sense here (or anywhere else for that matter.)

**Comment by celiot-IDM** [comment]
Note: if combining cadre results is OK, here's example code for how to do it:

```
resultsMerged <-
  pacehrh::RunExperiments(scenarioName = "MergedModel",
                          trials = 20)
extResultsMerged <- pacehrh::SaveExtendedSuiteResults(resultsMerged, run = "Run_M")
cadreAllocMerged <- pacehrh::SaveCadreAllocations(extResultsMerged)

resultsBasic <-
  pacehrh::RunExperiments(scenarioName = "BasicModel",
                          trials = 20)
extResultsBasic <- pacehrh::SaveExtendedSuiteResults(resultsBasic, run = "Run_B")
cadreAllocBasic <- pacehrh::SaveCadreAllocations(extResultsBasic)

resultsComprehensive <-
  pacehrh::RunExperiments(scenarioName = "ComprehensiveModel",
                          trials = 20)
extResultsComprehensive <- pacehrh::SaveExtendedSuiteResults(resultsComprehensive, run = "Run_C")
cadreAllocComprehensive <- pacehrh::SaveCadreAllocations(extResultsComprehensive)

SR <- data.table::rbindlist(list(extResultsMerged, extResultsBasic, extResultsComprehensive), fill=TRUE)
CA <- data.table::rbindlist(list(cadreAllocMerged, cadreAllocBasic, cadreAllocComprehensive), fill=TRUE)

CA[is.na(CA)] <- 0

stats <- pacehrh::ComputeSummaryStats(SR, CA)
```

> does pacehrh:::SaveCadreAllocations not work with the results from pacehrh::ReadAndCollateSuiteResults after this change? I ran into error: more than one scenario in suite results

No. That's the breaking change. It's safer - and a lot faster - to combine the suite results *after* they've been analyzed.

**Comment by rhanIDM** [comment]
@celiot-IDM Hi Charles, can you point me to where we currently store the annual overheads for all cadre members?

File: pacehrh/R/pace_cadre_process.R (hunk 1 above)
Comment (celiot-IDM) [reply pacehrh/R/pace_cadre_process.R:195]: > @celiot-IDM Hi Charles, can you point me to where we currently store the annual overheads for all cadre members?

@rhanIDM I was going to leave you to guess ... just kidding :-)

```
# Read overhead times and save to file
cadreOverheadTimes <- pacehrh::SaveCadreOverheadData(filepath = "_COD.csv")

# Read overhead times without saving to file
cadreOverheadTimes <- pacehrh::SaveCadreOverheadData()
```

I went with the practices from the original post-processing, so times are in HOURS, not minutes as they are reported elsewhere. We should decided if we want to move to hours everywhere.

**Comment by celiot-IDM** [comment]
@MeWu-IDM : could you please check that the build images are running the most recent version of testthat (should be 3.1.8). I added a unit test call to testthat::expect_no_message(), which was committed in September 2022, and broke the build.

**Comment by MeWu-IDM** [comment]
> @MeWu-IDM : could you please check that the build images are running the most recent version of testthat (should be 3.1.8). I added a unit test call to testthat::expect_no_message(), which was committed in September 2022, and broke the build.

This is likely because the build image needs update, it is passing with the new version
https://github.com/InstituteforDiseaseModeling/PACE-HRH/actions/runs/5085609827/jobs/9139943322
I also create a PR to allow update easily for the future, see https://github.com/InstituteforDiseaseModeling/PACE-HRH/pull/264

File: pacehrh/R/pace_cadre_process.R [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+              NumContactsPer)]
+
+  # Join Scenario information to the output
+  out <- dplyr::left_join(out,
+                          BVE$scenario[c("UniqueID",
+                                         "WeeksPerYr",
+                                         "HrsPerWeek",
+                                         "BaselinePop",
+                                         "DeliveryModel")],
+                          by = dplyr::join_by(Scenario_ID == UniqueID))
```
Comment (MeWu-IDM) [inline pacehrh/R/pace_cadre_process.R:132]: We should add this to force user to upgrade to the minimum version that has join_by
```
Imports:
    dplyr (>= 1.0.0),
```
@rhanIDM had encounter the issue which is as described as here https://r-pkgs.org/description.html#sec-description-imports-suggests-minium-version

 _If someone has an older package that doesn’t have a function your package needs, they’ll get an unhelpful error message if your package does not advertise the minimum version it needs. However, if you state a minimum version, they’ll automatically get an upgrade when they install your package._

File: pacehrh/R/pace_post_process.R
Comment (rhanIDM) [inline pacehrh/R/pace_post_process.R:1]: line 25 to line 37 contains "Overhead_staff" which we no longer use, should be removed

Comment (celiot-IDM) [reply pacehrh/R/pace_post_process.R:1]: I'm planning to remove the entire ReadAndCollate function as soon as I get confirmation that the new stuff works.

Comment (rhanIDM) [reply pacehrh/R/pace_post_process.R:1]: Thanks Charles. That makes a lot of sense.

File: pacehrh/R/pace_cadre_process.R (hunk 2 above)
Comment (BHagedorn-IDM) [reply pacehrh/R/pace_cadre_process.R:146]: @celiot-IDM  We cannot delete this, at least not as part of this pull request. We may eventually make the design decision that we want to remove the hours per week and weeks per year entirely, but that is a larger choice that involves removing scenarios columns etc. as well. I'm not convinced we want to do that at this stage. @MeWu-IDM if we want to move this elsewhere, please feel free to make that change as you see fit.

**Comment by rhanIDM** [comment]
@celiot-IDM, is RoleID required to be unique for task time allocation to cadre to compute correctly?

File: pacehrh/R/pace_cadre_process.R (hunk 2 above)
Comment (celiot-IDM) [reply pacehrh/R/pace_cadre_process.R:146]: Thank you @BHagedorn-IDM .

The logic for keeping the HPW_stochastic computation goes like this: the SaveExtendedSuiteResults() function is a "widening" function. In other words, it adds columns to the standard suite results. But this reasoning, HPW_stochastic is just another column.

The counter-argument is that we already have a place in the main code that does all the stochastic calculations, so why sneak another stochastic calculation in at the last moment. But this reasoning, the HPW_stochastic computation should be moved to be part of the RunExperiment() processing. Unfortunately, that's a bigger lift than just tacking it on at the end. So if we can live with this sub-optimal implementation for now, we should.

File: pacehrh/R/pace_cadre_process.R (hunk 3 above)
Comment (celiot-IDM) [reply pacehrh/R/pace_cadre_process.R:132]: Added minimum version check for dplyr.

File: pacehrh/R/pace_cadre_config.R [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+  # Abort if either read fails
+  if (is.null(dataRows) | is.null(headerRows)){
+    return(NULL)
+  }
+
+  # Remove the first two entries from the header rows to focus on the members and
+  # year info. Extract members and years, then combine into a single token to
+  # use in the subsequent table melt operation.
+  hr <- headerRows[-c(1,2)]
```
[the remark this answers is not in the document]
Comment (celiot-IDM) [reply pacehrh/R/pace_cadre_config.R:102]: Sanity check added.

File: pacehrh/R/pace_cadre_process.R (hunk 3 above)
Comment (celiot-IDM) [reply pacehrh/R/pace_cadre_process.R:132]: Added minimum version check for dplyr.

File: pacehrh/DESCRIPTION [hunk 5]
```diff
@@ -11,7 +11,7 @@ License: MIT + file LICENSE
 Imports:
     assertthat,
     data.table,
-    dplyr,
+    dplyr (>= 1.0.0),
```
Comment (MeWu-IDM) [inline pacehrh/DESCRIPTION:14]: it should be 1.1.0

**Comment by celiot-IDM** [comment]
> @celiot-IDM, is RoleID required to be unique for task time allocation to cadre to compute correctly?

The overhead times calculation will work just fine with duplicate RoleID values. After that it's up to you (the researcher) to collate RoleID's with the role information embedded in the headers of the cadre task sheets.

File: pacehrh/DESCRIPTION (hunk 5 above)
Comment (celiot-IDM) [reply pacehrh/DESCRIPTION]: Changed.

**Review by MeWu-IDM — APPROVED** [review]
We were able to update the scripts to produce the same cadre allocation results (Run_simulation.R is updated) 
Regression tests are also updated to cover the new cadre calculation.

**Review by rhanIDM — APPROVED** [review]
Run_simulations.R is updated to post-process new cadre allocations results structure. Model_inputs_demo.xlsx is updated to new structure and we were able to use these inputs to confirm the calculation results match the old results.

**Pull request merged**